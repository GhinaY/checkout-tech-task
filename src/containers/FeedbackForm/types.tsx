import { FieldError, ValidationRule } from 'react-hook-form';
import { FeedbackSubmission } from '../../types';
import { TextFieldProps } from '@mui/material/TextField';

export type ControlledTextFieldProps = TextFieldProps  & {
  name: keyof FeedbackSubmission,
  fieldError: FieldError | undefined,
  helperText?: string,
  required?: boolean,
  maxLength?: number,
  additionalRules?: {[key: string]: ValidationRule },
};
