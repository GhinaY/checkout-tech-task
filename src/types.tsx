export type FeedbackSubmission = {
  name: string,
  email: string,
  rating: number,
  comment: string,
};

export type RatingsDistribution = {
  [key: number]: number,
};
