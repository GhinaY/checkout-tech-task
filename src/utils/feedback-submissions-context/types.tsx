import { FeedbackSubmission, RatingsDistribution } from '../../types'

export type FeedbackSubmissionsContextType = {
  addNewSubmission: (newSubmission: FeedbackSubmission) => void,
  ratingsDistribution: RatingsDistribution,
  submissions: FeedbackSubmission[],
}
