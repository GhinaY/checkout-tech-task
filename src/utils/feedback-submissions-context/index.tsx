import React, { createContext, useState } from 'react';
import { FeedbackSubmission, RatingsDistribution } from '../../types';
import { FeedbackSubmissionsContextType } from './types';

export const FeedbackSubmissionsContext = createContext<FeedbackSubmissionsContextType>(null!);

const initialRatingsDistribution: RatingsDistribution = [1,2,3,4,5].reduce((obj, rating) => {
  return {
    ...obj,
    [rating]: 0,
  };
}, {});

interface Props {
  children: React.ReactNode;
}

export const FeedbackSubmissionsContextProvider: React.FC<Props> = ({ children }: Props) => {
  const [submissions, setSubmissions] = useState<FeedbackSubmission[]>([]);
  const [ratingsDistribution, setRatingsDistribution] = useState(initialRatingsDistribution);

  // adds the new submission to the list and increases the ratingsDistribution counters as needed
  const addNewSubmission = (newSubmission: FeedbackSubmission) => {
    // in a real production app, this would more realistically be an API call to save to a real database
    // for this purely-frontend app though, a context works fine
    setSubmissions(submissions.concat(newSubmission));

    // increase the counter for the rating submitted by 1
    const newRatingValue = newSubmission.rating;
    const newRatingsDistribution = {
      ...ratingsDistribution,
      [newRatingValue]: (ratingsDistribution[newRatingValue] ?? 0) + 1,
    }
    setRatingsDistribution(newRatingsDistribution)
  };

  return (
    <FeedbackSubmissionsContext.Provider value={{ addNewSubmission, ratingsDistribution, submissions }}>
      {children}
    </FeedbackSubmissionsContext.Provider>
  );
};
