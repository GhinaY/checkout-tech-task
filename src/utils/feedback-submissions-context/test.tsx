/* eslint-disable react-refresh/only-export-components */
import { useContext } from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import { FeedbackSubmissionsContext, FeedbackSubmissionsContextProvider } from './index'

const mockSubmission = {
  name: 'test name',
  email: 'test@test.com',
  rating: 5,
  comment: 'test comment',
}

const TestConsumer = () => {
  const { 
    addNewSubmission, 
    ratingsDistribution,
    submissions,
  } = useContext(FeedbackSubmissionsContext);

  return (
    <>
      Number of submissions: {submissions.length}
      {submissions.map((submission, i) => (
          <span key={i}>{submission.email}: {submission.comment}</span>
      ))}
      {Object.entries(ratingsDistribution).map(
        ([rating, count]) => (
            <span key={rating}>{rating}: {count}</span>
        )
      )}
      <button onClick={() => addNewSubmission(mockSubmission)}>
          Add new submission
      </button>
    </>
  )
}
describe('FeedbackSubmissionsContextProvider', () => {
  beforeEach(() => {
    render(
      <FeedbackSubmissionsContextProvider>
        <TestConsumer />
      </FeedbackSubmissionsContextProvider>
    );
  });

  test('Provides initial ratingsDistribution correctly', () => {  
    const text = screen.getByText(
      /1: 0/i
    );
    expect(text.textContent).toBeTruthy()
  })

  test('Provides initial submissions list correctly', () => {  
    const text = screen.getByText(
      /Number of submissions: 0/i
    );
    expect(text.textContent).toBeTruthy()
  })

  describe('addNewSubmission value', () => {
    test('updates ratingsDistribution values', async () => {  
      const initialText = screen.getByText(
        /5: 0/i
      );
      expect(initialText.textContent).toBeTruthy()

      fireEvent.click(screen.getByText(/Add new submission/i));

      const updatedText = screen.getByText(
        /5: 1/i
      );
      
      expect(updatedText.textContent).toBeTruthy();
    })

    test('updates submissions values', () => {  
      const initialListLengthText = screen.getByText(
        /Number of submissions: 0/i
      );

      expect(initialListLengthText.textContent).toBeTruthy()

      fireEvent.click(screen.getByText(/Add new submission/i));

      const updatedListLengthText = screen.getByText(
        /Number of submissions: 1/i
      );
      const updatedListValueText = screen.getByText(
        /test@test.com: test comment/i
      );

      expect(updatedListLengthText.textContent).toBeTruthy()
      expect(updatedListValueText.textContent).toBeTruthy()
    })
  });
});
