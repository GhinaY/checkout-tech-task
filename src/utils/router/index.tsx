import { Navigate, createBrowserRouter } from 'react-router-dom';
import FormPage from '../../pages/FormPage';
import ResultsPage from '../../pages/ResultsPage';

export const paths = [
  {
    path: '/',
    element: <FormPage />,
  },
  {
    path: '/results',
    element: <ResultsPage />,
  },
  { 
    path: '*', 
    element: <Navigate to='/' replace /> },
]

const router = createBrowserRouter(paths);

export default router;
