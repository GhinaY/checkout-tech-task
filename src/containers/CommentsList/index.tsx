import { useContext } from 'react';
import Stack from '@mui/material/Stack';
import Divider from '@mui/material/Divider';
import Typography from '@mui/material/Typography';
import Comment from '../../components/Comment';
import { FeedbackSubmissionsContext } from '../../utils/feedback-submissions-context';
import { SxProps } from '@mui/system';

export const PLACEHOLDER_TEXT = 'Hmm... Looks like there are no comments to show';

// Including SxProps allows the parent to style the container as needed
export type Props = SxProps & {
  commentCount?: number,
};

function CommentsList({ commentCount = 3, ...containerStylingProps }: Props) {
  const { submissions } = useContext(FeedbackSubmissionsContext);

  return (
    <Stack spacing={2} divider={<Divider variant='middle' sx={{...containerStylingProps}} />}>
      {/* ^ Column stack that wraps the comments 
      and dynamically adds dividers between each two comments */}

      {/* Added a placeholder message for visibility. 
      i.e. to keep the user informed of why there's no comments shown */}
      {submissions.length ? 
        submissions.slice(-commentCount).map((submission, i) => (
          <Comment {...submission} key={i}/>
        )) : (
        <Typography>
          {PLACEHOLDER_TEXT}
        </Typography>
      )}
    </Stack>
  );
}

export default CommentsList;
