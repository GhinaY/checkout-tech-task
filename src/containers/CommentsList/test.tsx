/* eslint-disable react-refresh/only-export-components */
import { render, screen } from '@testing-library/react'
import { FeedbackSubmissionsContext } from '../../utils/feedback-submissions-context';
import { FeedbackSubmission } from '../../types';
import CommentsList, { PLACEHOLDER_TEXT, Props } from './index'

vi.mock('../../components/Comment', () => {
  return {
    default: () => <div>test comment</div>
  }
})

const mockSubmission: FeedbackSubmission = {
  name: 'test name',
  email: 'test@test.com',
  rating: 5,
  comment: 'test comment',
}

describe('CommentList', () => {
  const mockSubmissionsList = Array(10).fill(mockSubmission);

  const renderComponent = (submissions: FeedbackSubmission[], componentProps?: Props) => {
    return render(
      <FeedbackSubmissionsContext.Provider value={{ 
        submissions, 
        ratingsDistribution: {}, 
        addNewSubmission: vi.fn() 
      }}>
        <CommentsList {...componentProps} />
      </FeedbackSubmissionsContext.Provider>
    );
  };

  test('Shows the number of comments specified', () => {
    renderComponent(mockSubmissionsList, {commentCount: 10});
    const result = screen.getAllByText(
      /test comment/i
    );
    expect(result.length).toBe(10);
  });

  test('Shows 3 comments by default if no number is specified', () => {
    renderComponent(mockSubmissionsList);
    const result = screen.getAllByText(
      /test comment/i
    );
    expect(result.length).toBe(3);
  });

  test('Matches screenshot', () => {
    const { asFragment } = renderComponent(mockSubmissionsList);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('If no submissions are available', () => {
    test('Does not render Comment components', () => {
      renderComponent([]);
      const result = screen.queryAllByText(
        /test comment/i
      );
      expect(result.length).toBe(0);
    });

    test('Shows placeholder text', () => {
      renderComponent([]);
      const result = screen.getAllByText(PLACEHOLDER_TEXT);
      
      expect(result.length).toBe(1);
    });
  })
});
