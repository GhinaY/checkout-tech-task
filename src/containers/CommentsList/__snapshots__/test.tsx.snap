// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentList > Matches screenshot 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-1sazv7p-MuiStack-root"
  >
    <div>
      test comment
    </div>
    <hr
      class="MuiDivider-root MuiDivider-middle css-ft4jij-MuiDivider-root"
    />
    <div>
      test comment
    </div>
    <hr
      class="MuiDivider-root MuiDivider-middle css-ft4jij-MuiDivider-root"
    />
    <div>
      test comment
    </div>
  </div>
</DocumentFragment>
`;
